File: backend/app/core/peeling_decoder.py
from collections import deque
from dataclasses import dataclass
from typing import Dict, List, Set

from app.core.fountain import degree_and_block_indices, robust_soliton_probabilities, xor_bytes


@dataclass
class PendingSymbol:
    unknown_indices: Set[int]
    data: bytes


class PeelingDecoder:
    def __init__(self, k: int, block_size: int, c: float = 0.03, delta: float = 0.05):
        self.k = k
        self.block_size = block_size
        self.probs = robust_soliton_probabilities(k, c=c, delta=delta)
        self.known_blocks: Dict[int, bytes] = {}
        self.pending: List[PendingSymbol] = []
        self.seen_seeds: Set[int] = set()

    @property
    def progress(self) -> int:
        return len(self.known_blocks)

    @property
    def is_complete(self) -> bool:
        return len(self.known_blocks) == self.k

    def add_symbol(self, seed: int, data: bytes) -> bool:
        if seed in self.seen_seeds:
            return False
        self.seen_seeds.add(seed)
        if self.is_complete:
            return True

        _, block_indices = degree_and_block_indices(seed, self.k, self.probs)
        unknown = set(block_indices)
        reduced = data
        for idx in block_indices:
            if idx in self.known_blocks:
                reduced = xor_bytes([reduced, self.known_blocks[idx]])
                unknown.discard(idx)

        if not unknown:
            return True
        if len(unknown) == 1:
            self._resolve(next(iter(unknown)), reduced)
        else:
            self.pending.append(PendingSymbol(unknown_indices=unknown, data=reduced))
        return True

    def _resolve(self, idx: int, data: bytes) -> None:
        if idx in self.known_blocks:
            return
        self.known_blocks[idx] = data
        queue = deque([idx])
        while queue:
            known_idx = queue.popleft()
            known_data = self.known_blocks[known_idx]
            still_pending: List[PendingSymbol] = []
            for sym in self.pending:
                if known_idx in sym.unknown_indices:
                    sym.data = xor_bytes([sym.data, known_data])
                    sym.unknown_indices = sym.unknown_indices - {known_idx}
                if len(sym.unknown_indices) == 0:
                    continue
                elif len(sym.unknown_indices) == 1:
                    new_idx = next(iter(sym.unknown_indices))
                    if new_idx not in self.known_blocks:
                        self.known_blocks[new_idx] = sym.data
                        queue.append(new_idx)
                else:
                    still_pending.append(sym)
            self.pending = still_pending

    def get_reconstructed_blocks(self) -> List[bytes]:
        if not self.is_complete:
            raise RuntimeError(f"Cannot reconstruct yet: only {self.progress}/{self.k} blocks known")
        return [self.known_blocks[i] for i in range(self.k)]

File: frontend/src/lib/peelingDecoder.js
/**
 * peelingDecoder.js
 *
 * JavaScript port of backend/app/core/peeling_decoder.py.
 *
 * Implements belief-propagation graph peeling to reconstruct a file from
 * a random, unordered, lossy stream of fountain symbols.
 *
 * Imports degreeAndBlockIndices from fountain.js — the same single source of
 * truth used by FountainEncoder. This is the critical link that makes the
 * decoder and encoder agree on which blocks each seed represents.
 */

import { degreeAndBlockIndices, robustSolitonProbabilities, xorBytes } from './fountain.js'

// ── Internal type: a symbol that's still awaiting resolution ──────────────────
//
// unknownIndices: Set<number> — block indices not yet resolved
// data:           Uint8Array  — XOR residue after known blocks are stripped out

// ── PeelingDecoder ────────────────────────────────────────────────────────────

/**
 * PeelingDecoder
 *
 * Port of peeling_decoder.py's PeelingDecoder class.
 *
 * Usage:
 *   const decoder = new PeelingDecoder(k, blockSize, 0.03, 0.05)
 *   decoder.addSymbol(seed, payload)  // call for each scanned QR frame
 *   decoder.isComplete                // true when all k blocks recovered
 *   decoder.getReconstructedBlocks()  // returns ordered Uint8Array[]
 */
export class PeelingDecoder {
  /**
   * @param {number} k         - Total number of source blocks (from metadata frame).
   * @param {number} blockSize - Size of each block in bytes (from metadata frame).
   * @param {number} c         - Robust Soliton c (default 0.03).
   * @param {number} delta     - Robust Soliton delta (default 0.05).
   */
  constructor(k, blockSize, c = 0.03, delta = 0.05) {
    this.k = k
    this.blockSize = blockSize
    this.probs = robustSolitonProbabilities(k, c, delta)

    /** @type {Map<number, Uint8Array>} index → reconstructed block bytes */
    this.knownBlocks = new Map()

    /** @type {Array<{ unknownIndices: Set<number>, data: Uint8Array }>} */
    this.pending = []

    /** @type {Set<number>} seeds already processed — deduplication */
    this.seenSeeds = new Set()
  }

  /** Number of blocks resolved so far. */
  get progress() {
    return this.knownBlocks.size
  }

  /** True once all k blocks have been recovered. */
  get isComplete() {
    return this.knownBlocks.size === this.k
  }

  /**
   * addSymbol(seed, data) → boolean
   *
   * Ingests one decoded QR payload. Returns false if the seed was a duplicate
   * (already seen), true otherwise.
   *
   * @param {number}     seed - Symbol seed from the packet header.
   * @param {Uint8Array} data - XOR payload bytes from the QR frame.
   * @returns {boolean}
   */
  addSymbol(seed, data) {
    if (this.seenSeeds.has(seed)) return false
    this.seenSeeds.add(seed)
    if (this.isComplete) return true

    const { blockIndices } = degreeAndBlockIndices(seed, this.k, this.probs)
    const unknownIndices = new Set(blockIndices)
    let reduced = data.slice()  // work on a copy; never mutate the caller's buffer

    // Strip out any blocks we already know.
    for (const idx of blockIndices) {
      if (this.knownBlocks.has(idx)) {
        reduced = xorBytes([reduced, this.knownBlocks.get(idx)])
        unknownIndices.delete(idx)
      }
    }

    if (unknownIndices.size === 0) {
      // All blocks in this symbol were already known — nothing new.
      return true
    }

    if (unknownIndices.size === 1) {
      // Degree-1: we can immediately resolve this block.
      this._resolve(next(unknownIndices), reduced)
    } else {
      // Still ambiguous — buffer for later.
      this.pending.push({ unknownIndices, data: reduced })
    }

    return true
  }

  /**
   * _resolve(idx, data) → void
   *
   * Records a newly discovered block and cascades: XORs it out of every
   * pending symbol that referenced it, potentially triggering further
   * resolutions in a BFS queue (the "peeling cascade").
   *
   * @param {number}     idx  - Block index being resolved.
   * @param {Uint8Array} data - The block's actual byte content.
   */
  _resolve(idx, data) {
    if (this.knownBlocks.has(idx)) return
    this.knownBlocks.set(idx, data)

    // BFS queue of newly resolved block indices to cascade through pending.
    const queue = [idx]

    while (queue.length > 0) {
      const knownIdx = queue.shift()
      const knownData = this.knownBlocks.get(knownIdx)
      const stillPending = []

      for (const sym of this.pending) {
        if (sym.unknownIndices.has(knownIdx)) {
          sym.data = xorBytes([sym.data, knownData])
          sym.unknownIndices.delete(knownIdx)
        }

        if (sym.unknownIndices.size === 0) {
          // All blocks resolved — symbol fully consumed.
          continue
        } else if (sym.unknownIndices.size === 1) {
          // Cascade: this symbol just became degree-1.
          const newIdx = next(sym.unknownIndices)
          if (!this.knownBlocks.has(newIdx)) {
            this.knownBlocks.set(newIdx, sym.data)
            queue.push(newIdx)
          }
        } else {
          stillPending.push(sym)
        }
      }

      this.pending = stillPending
    }
  }

  /**
   * getReconstructedBlocks() → Uint8Array[]
   *
   * Returns the ordered array of reconstructed blocks once isComplete is true.
   * Pass the result directly to chunker.js reconstruct().
   *
   * @throws {Error} If called before isComplete.
   * @returns {Uint8Array[]}
   */
  getReconstructedBlocks() {
    if (!this.isComplete) {
      throw new Error(
        `Cannot reconstruct yet: only ${this.progress}/${this.k} blocks known`
      )
    }
    return Array.from({ length: this.k }, (_, i) => this.knownBlocks.get(i))
  }
}

// ── Internal helper ───────────────────────────────────────────────────────────

/** Returns the single element of a size-1 Set. */
function next(set) {
  return set.values().next().value
}
